### PR TITLE
prov/verbs: defer uffd events

### PR DESCRIPTION
**Pull request by jswaro**
closes ofiwg/libfabric#5554

WIP

### CODE REVIEW

**Comment by wckzhang** [comment]
Nit: prov/verbs <-- there's nothing explicitly verbs related in this patch

**Comment by jswaro** [comment]
> Nit: prov/verbs <-- there's nothing explicitly verbs related in this patch

Agreed. Not sure what I was thinking.

**Comment by shefty** [comment]
I will create a patch to update the rbtree.  IMO, that needs to come first, so that we can remove entries from the tree without free() being called.

By adding a 'stale' flag, it would need to be included as part of the search criteria if the rbtree isn't updated.

File: prov/util/src/util_mr_cache.c [hunk 1]
```diff
@@ -121,9 +150,11 @@ void ofi_mr_cache_notify(struct ofi_mr_cache *cache, const void *addr, size_t le
 	iov.iov_base = (void *) addr;
 	iov.iov_len = len;
 
+	pthread_mutex_lock(&cache->stale_lock);
 	for (entry = cache->storage.overlap(&cache->storage, &iov); entry;
 	     entry = cache->storage.overlap(&cache->storage, &iov))
```
Comment (shefty) [inline prov/util/src/util_mr_cache.c]: The problem with this change is that the overlap find operation does not include the stale check as part of its search.

File: prov/util/src/util_mr_cache.c [hunk 2]
```diff
@@ -149,6 +181,7 @@ bool ofi_mr_cache_flush(struct ofi_mr_cache *cache)
 	bool empty;
 
 	pthread_mutex_lock(&cache->monitor->lock);
+	util_mr_process_stale_entries(cache);
```
Comment (shefty) [inline prov/util/src/util_mr_cache.c]: We can't hold the monitor lock during any entry cleanup.  That can still result in a hang.

File: prov/util/src/util_mr_cache.c [hunk 3]
```diff
@@ -149,6 +181,7 @@ bool ofi_mr_cache_flush(struct ofi_mr_cache *cache)
...
 	empty = mr_cache_flush(cache);
```
Comment (shefty) [inline prov/util/src/util_mr_cache.c]: It also means that the existing flush code must change.

File: prov/util/src/util_mr_cache.c (hunk 1 above)
Comment (jswaro) [reply prov/util/src/util_mr_cache.c]: Still not quite done with that yet. I think we can avoid the stale check if we prune the entry in the first place.

File: prov/util/src/util_mr_cache.c (hunk 2 above)
Comment (jswaro) [reply prov/util/src/util_mr_cache.c]: I've moved the code outside of the lock again.

**Comment by jswaro** [comment]
@shefty How did you want to coordinate on this? I see more changes in your other PR, but I had assumed both PRs were attempting to fix the same problem. I can try to pull your changes into my local branch, but I'd like to get on the same page in terms of changes. Is your PR only addressing the rbtree, or is it handling more than that?

**Comment by shefty** [comment]
Both PRs are eventually getting to the same problem.  The approach I took was to fix the current code to handle flushing the lru_list outside of holding the monitor lock.  The lru_list cleanup has the same problem, and fixing it requires significant code restructuring.  That's what I have so far.

With those changes in place, adding a flush_list and handling that cleanup in the same fashion is relatively straightforward.  I actually have another patch I'm working on over these changes to add that, but I just started on it.

What I haven't done is try to move the registration code outside of the monitor lock.  That's more difficult and requires asynchronous operation...  But I'm also not convinced that it's likely to be needed.

**Comment by shefty** [comment]
Btw, read 'flush_list' above as the same as your 'stale_list'.  (I couldn't remember the name.)  My PR captures the same cleanup as yours, but focuses on the existing lru_list.

**Comment by shefty** [comment]
I've updated #5562 with what I think are the full set of changes to move memory deregistration and free calls to outside of the monitor lock.  This includes MR entries for which we receive notification (from the uffd callback), and other paths that might result in flushing the cache (such as flushing the lru list).  I haven't tested the changes yet, but they should be good for reviewing.

**Pull request closed without merging**